### DISCUSSION TITLE
How to place tests in a monorepo / how to decide the right structure

### DISCUSSION THREAD

**Discussion by TomFreudenberg** [vercel/turborepo#2320] [Help] [answered]
Which division makes sense or is recommended

There are 3 applications in a monorepo:

1. Frontend App (Nextjs) app.my.domain
2. Backend Api (TRPC Service) api.my.domain
3. Authentication Site auth.my.domain

There are several unit and end-2-end tests. 

80% of jest tests are "private" to their package (app, api, auth). But nearly all of the e2e tests need content (source) from mostly 2 or all 3 services to run.

Would you consider to place at least the e2e tests (and maybe also the 20% common jest tests) into a separated package or app?

The structure as currently decided:

```
my-repo
- apps
  > frontend
  > api
  > auth
  > storybook
  > docsify
  > e2e-tests
  > tests (common)
```

Thanks for any feedback or guiding.

Tom

P.S.: Is there any objective decision rule to NOT place the tests into a separated package?

**Comment by anthonyshew** [comment, marked as answer]
Generally, if tests are only concerned with one package, it's a good idea to put them in that package.

E2Es can be coupled to into the package of the application they are testing or made into their own packages. From what I've seen, it is more common to make E2Es their own package - but, as always, it depends!
